FSType is now optional in Volume objects and 'ext4' used when it's when omitted.

jsafrane opened the pull request:
This is follow up of https://github.com/kubernetes/kubernetes/pull/17650#discussion_r45755077.

Most volume plugins use `SafeFormatAndMount`, which uses ext4 by default.

`FlexVolume` plugin has `FSType` attribute 'omitempty', so reflect it in the
description of the type.

Also, update AWS EBS plugin to have the same style as the rest of plugins - see https://github.com/kubernetes/kubernetes/pull/17650#discussion_r45618690

@swagiaal, would you mind a short review?

swagiaal commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -110,7 +110,7 @@ func (plugin *awsElasticBlockStorePlugin) newBuilderInternal(spec *volume.Spec,
 		partition:   partition,
 		readOnly:    readOnly,
-		diskMounter: &mount.SafeFormatAndMount{plugin.host.GetMounter(), exec.New()}}, nil
+		diskMounter: &mount.SafeFormatAndMount{mounter, exec.New()}}, nil
```
Could you separate this change into its own PR it is not related to this PR.

swagiaal commented:
@jsafrane thanks for following up...
The validation changes look good to me.
The api doc changes look good to but you will have to do them in v1/types.go as well.
The aws ebs change looks good but it should be in a separate PR

jsafrane commented:
> you will have to do them in v1/types.go as well.

Will do shortly.

> The aws ebs change looks good but it should be in a separate PR

Separate commit is not enough?

swagiaal commented:
> Separate commit is not enough?

I think it should be a separate PR. Separate commits are for related sequential changes building up to one big modification.

jsafrane commented:
Ok, I removed AWS changes from the PR. I guess we'll sort it out when we rewrite plugins with the new Attach interface.

swagiaal commented:
Hmm.. the aws change seems to still be there
Doc changes look good

jsafrane commented:
Weird, I would swear I removed the patch. Now it's really removed.

swagiaal commented:
@jsafrane thanks... LGTM

markturansky commented on pkg/api/types.go:
```diff
@@ -502,9 +502,9 @@ type FCVolumeSource struct {
+	// Filesystem type to mount.
 	// Must be a filesystem type supported by the host operating system.
-	// Ex. "ext4", "xfs", "ntfs"
+	// Ex. "ext4", "xfs", "ntfs". The default one depends on FlexVolume script.
```
"The default filesystem depends on ..."

jsafrane replied:
Updated, thanks.

markturansky commented on pkg/api/v1/types.go:
```diff
@@ -663,9 +663,9 @@ type GCEPersistentDiskVolumeSource struct {
+	// Filesystem type to mount.
 	// Must be a filesystem type supported by the host operating system.
-	// Ex. "ext4", "xfs", "ntfs"
+	// Ex. "ext4", "xfs", "ntfs". The default one depends on FlexVolume script.
```
"The default filesystem depends on ..."

jsafrane replied:
Updated, thanks.

markturansky commented:
LGTM after the wording changes above.

@bgrant0607 please review for API changes?

bgrant0607 commented:
This isn't strictly backward compatible. What's the motivation?

markturansky commented:
None of these API fields are strictly required as all plugins involved in this PR use SafeFormatAndMount which defaults to "ext4".

Jan is looking to make the API reflect the reality of the code.

swagiaal commented:
In addition all the fields say "omitempty"

jsafrane commented:
@bgrant0607, as Mark wrote, motivation is to unify all the volumes. All of them already use the same code on the lower levels to format and mount a block device, this patch propagates this change to higher levels, i.e. type definition (add `omitempty` where applicable), validation and documentation.

jsafrane commented:
Rebased and updated description according to @markturansky suggestions.

jsafrane commented:
Rebased

bgrant0607 commented:
cc @thockin

bgrant0607 commented:
@markturansky The field is not set by default AFAICT:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/defaults.go

markturansky commented:
@bgrant0607 the default value is found in the mount utils. Perhaps the wrong place, but you'll find it here:  https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_linux.go#L279

bgrant0607 commented:
@markturansky You're right. Not the right place. That won't be reflected in the API. When we say a field has a "default value", that value is represented explicitly when the user does a GET on the resource.

bgrant0607 commented:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#defaulting

bgrant0607 commented:
If the omitempty changes reflect the code, it's better to make them consistent, but we'll need a release note to warn users about the incompatible API spec change, which may affect auto-generated clients.

The comments can't say "Defaults to...", because that's not true. It's more "Implicitly inferred to be...".

Please file an issue to fix this to comply with standard defaulting behavior in a future API version.

jsafrane commented:
Rebased and comments updated to `Implicitly inferred to be "ext4"`. How do I update release notes?

swagiaal commented:
> Please file an issue to fix this to comply with standard defaulting behavior in a future API version.

@bgrant0607 I'll take care of this: https://github.com/kubernetes/kubernetes/issues/20332

bgrant0607 commented:
Re. release note: I labeled this PR appropriately. Please draft a release note and put it in the PR description.

bgrant0607 commented:
Please add "if unspecified"

jsafrane commented:
> Please add "if unspecified"

Added, now the comment says: `Implicitly inferred to be "ext4" if unspecified.`

jsafrane committed:
Remove validation of *VolumeSource.FSType.

Most volume plugins use SafeFormatAndMount, which uses ext4 by default.

FlexVolume plugin has FSType attribute 'omitempty', so reflect it in the
description of the type.

jsafrane commented:
rebased

bgrant0607 commented:
LGTM
